odoo/tools/sql.py

```diff
@@ -23,23 +23,108 @@
     'SET DEFAULT': 'd',
 }
 
+
+class SQL:
+    """ An object that wraps an SQL string with its positional parameters.
+    The positional parameters can be either actual parameters or SQL objects
+    themselves. In the latter case, those SQL objects are merged into the main
+    SQL object, and their respective parameters are merged at the right position.
+    """
+    __slots__ = ['__string', '__args']
+
+    def __init__(self, string: str = "", *args):
+        if isinstance(string, SQL):
+            string, args = string.__string, string.__args
+        string % tuple("x" for arg in args)
```

Comment:
I introduced this enforcement because I was almost unable to debug a SQL object with a missing parameter.  The issue is that you generally have a _composition_ of SQL objects, finding which one misses its parameter and where it has been created is really non-trivial.

For this problem, I suggest to add a _parameter placeholder_, i.e., an object that contains a value which may be updated post-hoc, and implements the SQL adaptation protocol.  So you would put that as a parameter anywhere in your SQL objects, and it would simply end up in the `sql.params` list.  If you set its internal value before executing that SQL object, that value will be adapted by `psycopg2`.